cobc/tree.h

```diff
@@ -1947,6 +1950,9 @@ struct cb_program {
 	unsigned int	flag_void		: 1;	/* void return for subprogram */
 	unsigned int	flag_decimal_comp	: 1;	/* program group has decimal computations */
 	unsigned int	flag_prototype		: 1;	/* Is a prototype */
+	unsigned int	flag_inspect_used		: 1;
+	unsigned int	flag_string_used		: 1;
+	unsigned int	flag_unstring_used		: 1;
```

Comment:
What do you think about dropping those three completely and build `inspect_st` and friends only when used the first time? This way yo have no change to tree.c at all and in codegen check the variable that you output.

Reply:
If I remove my changes to `tree.{c.h}`, how am I supposed to access strings states in other functions other than `_init` ones (where I build those states)? And how can I tell in `codegen.c` that those states have been built?

Reply:
I suggest to only drop the flags; then you only have the tree in tree.c, with default initialization to binary zero = NULL.
instead of

```c
flag_unstring_used = 1;
```

you'd then do

```c
if (!unstring_st) unstring_st = cb_build_direct ("unstring_st", 0);
```

Reply:
I tried something. I had to place the build of string states in the parser directly for it to work.

Reply:
As I now see the changed code I do think that's nicer.

> I had to place the build of string states in the parser directly for it to work.

But now I'm confused - why doesn't that work in the old place (typeck.c)?

Reply:
As far as I understand, there are some string-related function calls in `parser.y` so the string states will not be initialized for them as they are initialized later in `typeck.c`. I could mix initialization in both files but I thought it was clearer to just initialize in `parser.y`.

Reply:
'I'm fine either way, now we only need to have that memcheck report improved.